### PR TITLE
[KubeRay][Doc] KubeRay doc migration (1/n)

### PR DESCRIPTION
**Pull request by kevin85421**
## Related issue number

I failed to create a child page under "Managed Kubernetes Services." I believe there might be some layer limitations in our documentation. The [VM launcher](https://github.com/ray-project/ray/blob/dea8ad7cb7dd98eb5a0d75a053847ce282f84230/doc/source/_toc.yml#L346-L351) also tried to do similar things, but I can't see an entry like "Deploying on YARN" in the sidebar.

Screen Shot 2023-08-21 at 4 06 26 PM

wants to add the doc in a deeper layer, but  

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by kevin85421** [comment]
cc @angelinalg would you mind reviewing this PR? Btw, I am not familiar with Ray docs. Are there any tests that I need to run in advance locally? Thanks!

**Review by architkulkarni — APPROVED** [review]
Stamping in case `cluster` codeowner approval is required.

File: doc/source/cluster/kubernetes/examples/mobilenet-rayservice.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+curl -LO https://raw.githubusercontent.com/ray-project/kuberay/master/ray-operator/config/samples/ray-service.mobilenet.yaml
+
+# Step 3.2: Create a RayService
+kubectl apply -f ray-service.mobilenet.yaml
+```
+
+* The [mobilenet.py](https://github.com/ray-project/serve_config_examples/blob/master/mobilenet/mobilenet.py) file requires `tensorflow` as a dependency. Hence, the YAML file uses `rayproject/ray-ml:2.5.0` instead of `rayproject/ray:2.5.0`.
+* `python-multipart` is required for the request parsing function `starlette.requests.form()`, so the YAML file includes `python-multipart` in the runtime environment.
+
+## Step 4: Forward the port of Serve
```
Comment (angelinalg) [inline doc/source/cluster/kubernetes/examples/mobilenet-rayservice.md:31]:
```suggestion
## Step 4: Forward the port for Ray Serve
```

File: doc/source/cluster/kubernetes/examples/mobilenet-rayservice.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+```sh
+kubectl port-forward svc/rayservice-mobilenet-serve-svc 8000
+```
+
+Note that the Serve service will be created after the Serve applications are ready and running. This process may take approximately 1 minute after all Pods in the RayCluster are running.
```
Comment (angelinalg) [inline doc/source/cluster/kubernetes/examples/mobilenet-rayservice.md:37]:
```suggestion
Note that the Ray Serve service is created after the Ray Serve applications are ready and running. This process may take approximately 1 minute after all Pods in the RayCluster are running.
```

File: doc/source/cluster/kubernetes/examples/mobilenet-rayservice.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+## Step 2: Install KubeRay operator
+
+Follow [this document](kuberay-operator-deploy) to install the latest stable KubeRay operator via Helm repository.
+Please note that the YAML file in this example uses `serveConfigV2`, which is supported starting from KubeRay v0.6.0.
+
+## Step 3: Install a RayService
+
+```sh
+# Step 3.1: Download `ray-service.mobilenet.yaml`
```
Comment (angelinalg) [inline doc/source/cluster/kubernetes/examples/mobilenet-rayservice.md:21]:
```suggestion

```

File: doc/source/cluster/kubernetes/examples/mobilenet-rayservice.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+kind create cluster --image=kindest/node:v1.23.0
+```
+
+## Step 2: Install KubeRay operator
+
+Follow [this document](kuberay-operator-deploy) to install the latest stable KubeRay operator via Helm repository.
+Please note that the YAML file in this example uses `serveConfigV2`, which is supported starting from KubeRay v0.6.0.
+
+## Step 3: Install a RayService
+
```
Comment (angelinalg) [inline doc/source/cluster/kubernetes/examples/mobilenet-rayservice.md:19]:
```suggestion

* Step 3.1: Download `ray-service.mobilenet.yaml`
```

File: doc/source/cluster/kubernetes/examples/mobilenet-rayservice.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+```sh
+# Step 3.1: Download `ray-service.mobilenet.yaml`
+curl -LO https://raw.githubusercontent.com/ray-project/kuberay/master/ray-operator/config/samples/ray-service.mobilenet.yaml
+
+# Step 3.2: Create a RayService
```
Comment (angelinalg) [inline doc/source/cluster/kubernetes/examples/mobilenet-rayservice.md:24]:
```suggestion
* Step 3.2: Create a RayService
```sh
```

File: doc/source/cluster/kubernetes/examples/mobilenet-rayservice.md [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+## Step 2: Install KubeRay operator
+
+Follow [this document](kuberay-operator-deploy) to install the latest stable KubeRay operator via Helm repository.
+Please note that the YAML file in this example uses `serveConfigV2`, which is supported starting from KubeRay v0.6.0.
+
+## Step 3: Install a RayService
+
+```sh
+# Step 3.1: Download `ray-service.mobilenet.yaml`
+curl -LO https://raw.githubusercontent.com/ray-project/kuberay/master/ray-operator/config/samples/ray-service.mobilenet.yaml
```
Comment (angelinalg) [inline doc/source/cluster/kubernetes/examples/mobilenet-rayservice.md:22]:
```suggestion
curl -LO https://raw.githubusercontent.com/ray-project/kuberay/master/ray-operator/config/samples/ray-service.mobilenet.yaml
```
add ```

File: doc/source/cluster/kubernetes/getting-started/raycluster-quick-start.md [hunk 7]
```diff
@@ -0,0 +1,153 @@
+(kuberay-raycluster-quickstart)=
+
+# RayCluster
+
+In this guide, we show you how to manage and interact with Ray clusters on Kubernetes.
+
+## Preparation
+
+* Install [kubectl](https://kubernetes.io/docs/tasks/tools/#kubectl) (>= 1.19), [Helm](https://helm.sh/docs/intro/install/) (>= v3.4), and [kind](https://kind.sigs.k8s.io/docs/user/quick-start/#installation).
+* Please make sure your Kubernetes cluster has at least 4 CPU and 4 GB RAM.
```
Comment (angelinalg) [inline doc/source/cluster/kubernetes/getting-started/raycluster-quick-start.md:10]:
```suggestion
* Make sure your Kubernetes cluster has at least 4 CPU and 4 GB RAM.
```

File: doc/source/cluster/kubernetes/getting-started/rayjob-quick-start.md [hunk 8]
```diff
@@ -0,0 +1,143 @@
+(kuberay-rayjob-quickstart)=
+
+# RayJob
+
+> Note: This is the alpha version of Ray Job Support in KubeRay. There will be ongoing improvements for Ray Job in the future releases.
```
Comment (angelinalg) [inline doc/source/cluster/kubernetes/getting-started/rayjob-quick-start.md:5]:
```suggestion
> Note: Ray Job Support in KubeRay is in alpha. Ray Job improvements are planned for future releases.
```

File: doc/source/cluster/kubernetes/getting-started/rayjob-quick-start.md [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+
+## Prerequisites
+
+* Ray 1.10 or higher
+* KubeRay v0.3.0+. (v0.6.0+ is recommended)
+
+## What is a RayJob?
+
+A RayJob manages 2 things:
```
Comment (angelinalg) [inline doc/source/cluster/kubernetes/getting-started/rayjob-quick-start.md:14]:
```suggestion
A RayJob manages two aspects:
```

File: doc/source/cluster/kubernetes/getting-started/rayservice-quick-start.md [hunk 10]
```diff
@@ -0,0 +1,293 @@
+(kuberay-rayservice-quickstart)=
+
```
Comment (angelinalg) [inline doc/source/cluster/kubernetes/getting-started/rayservice-quick-start.md:2]:
```suggestion
This guide focuses solely on the Ray Serve multi-application API, which is available starting from Ray version 2.4.0.
```

File: doc/source/cluster/kubernetes/getting-started/rayservice-quick-start.md [hunk 11]
```diff
@@ -0,0 +1,293 @@
...
+# RayService
+
+## Prerequisites
+
+This guide focuses solely on the Ray Serve multi-application API, which is available starting from Ray version 2.4.0.
```
Comment (angelinalg) [inline doc/source/cluster/kubernetes/getting-started/rayservice-quick-start.md:7]:
```suggestion
```

File: doc/source/cluster/kubernetes/getting-started/rayservice-quick-start.md [hunk 12]
```diff
@@ -0,0 +1,293 @@
...
+
+* Ray 2.4.0 or newer.
+* KubeRay 0.6.0, KubeRay nightly, or newer.
+
+## What is a RayService?
+
+A RayService manages 2 things:
```
Comment (angelinalg) [inline doc/source/cluster/kubernetes/getting-started/rayservice-quick-start.md:14]:
```suggestion
A RayService manages 2 aspects:
```

File: doc/source/cluster/kubernetes/getting-started/rayjob-quick-start.md (hunk 8 above)
Comment (angelinalg) [reply doc/source/cluster/kubernetes/getting-started/rayjob-quick-start.md:5]: Suggest using the {warning} component like the one on the landing page of Ray Serve (which has  {tip} component).

File: doc/source/cluster/kubernetes/index.md [hunk 13]
```diff
@@ -19,12 +19,14 @@ different Ray versions in the same Kubernetes cluster.
   Find source document here: https://docs.google.com/drawings/d/1E3FQgWWLuj8y2zPdKXjoWKrfwgYXw6RV_FWRwK8dVlg/edit
 ```
 
+KubeRay introduces three distinct Kubernetes Custom Resource Definitions (CRDs): **RayCluster**, **RayJob**, and **RayService**.
+These are designed to assist users in efficiently managing Ray clusters tailored to various use cases.
```
Comment (angelinalg) [inline doc/source/cluster/kubernetes/index.md:23]:
```suggestion
These CRDs assist users in efficiently managing Ray clusters tailored to various use cases.
```

File: doc/source/cluster/kubernetes/index.md [hunk 14]
```diff
@@ -19,12 +19,14 @@ different Ray versions in the same Kubernetes cluster.
...
 
-Concretely, you will learn how to:
+Check out the [Getting Started](kuberay-quickstart) page to learn the basics of KubeRay and follow the quick start guides to run your first Ray application on Kubernetes with KubeRay!
```
Comment (angelinalg) [inline doc/source/cluster/kubernetes/index.md:25]:
```suggestion
See [Getting Started](kuberay-quickstart) to learn the basics of KubeRay and follow the quickstart guides to run your first Ray application on Kubernetes with KubeRay.
```

File: doc/source/cluster/kubernetes/index.md [hunk 15]
```diff
@@ -64,55 +81,55 @@ The Ray docs present all the information you need to start running Ray workloads
...
-        
+
     .. grid-item-card::
 
-        **User Guides**
+        **Ecosystem**
         ^^^
     
-        Learn best practices for configuring Ray clusters on Kubernetes.
+        Integrate KubeRay with 3rd party Kubernetes ecosystem tools.
```
Comment (angelinalg) [inline doc/source/cluster/kubernetes/index.md:90]:
```suggestion
        Integrate KubeRay with third party Kubernetes ecosystem tools.
```

File: doc/source/cluster/kubernetes/index.md [hunk 16]
```diff
@@ -64,55 +81,55 @@ The Ray docs present all the information you need to start running Ray workloads
...
+    
+        Check the KubeRay benchmark results.
+    
+        +++
+        .. button-ref:: kuberay-benchmarks
+            :color: primary
+            :outline:
+            :expand:
+
+            Check the benchmark results
```
Comment (angelinalg) [inline doc/source/cluster/kubernetes/index.md:113]:
```suggestion
            Benchmark results
```

File: doc/source/cluster/kubernetes/index.md [hunk 17]
```diff
@@ -64,55 +81,55 @@ The Ray docs present all the information you need to start running Ray workloads
...
     
     .. grid-item-card::
 
-        **API Reference**
+        **Troubleshooting**
         ^^^
     
-        Find API references on RayCluster configuration.
+        Check the KubeRay troubleshooting guides.
```
Comment (angelinalg) [inline doc/source/cluster/kubernetes/index.md:120]:
```suggestion
        Consult the KubeRay troubleshooting guides.
```

File: doc/source/cluster/kubernetes/index.md [hunk 18]
```diff
@@ -64,55 +81,55 @@ The Ray docs present all the information you need to start running Ray workloads
...
     
         +++
-        .. button-ref:: kuberay-api-reference
+        .. button-ref:: kuberay-troubleshooting
             :color: primary
             :outline:
             :expand:
 
-            Check API references
+            Check the troubleshooting guides
```
Comment (angelinalg) [inline doc/source/cluster/kubernetes/index.md:128]:
```suggestion
            Troubleshooting guides
```

File: doc/source/cluster/kubernetes/index.md [hunk 19]
```diff
@@ -64,55 +81,55 @@ The Ray docs present all the information you need to start running Ray workloads
...
     
         +++
-        .. button-ref:: kuberay-guides
+        .. button-ref:: kuberay-ecosystem-integration
             :color: primary
             :outline:
             :expand:
 
-            Read the User Guides
+            Explore Kubernetes ecosystem
```
Comment (angelinalg) [inline doc/source/cluster/kubernetes/index.md:98]:
```suggestion
            Ecosystem Guides
```

File: doc/source/cluster/kubernetes/k8s-ecosystem/ingress.md [hunk 20]
```diff
@@ -0,0 +1,132 @@
+(kuberay-ingress)=
+
+# Ingress
+
+Here we provide some examples to show how to use ingress to access your Ray cluster.
```
Comment (angelinalg) [inline doc/source/cluster/kubernetes/k8s-ecosystem/ingress.md:5]:
```suggestion
Two examples show how to use ingress to access your Ray cluster:
```

File: doc/source/cluster/kubernetes/k8s-ecosystem/ingress.md [hunk 21]
```diff
@@ -0,0 +1,132 @@
...
+
+  * [Example: AWS Application Load Balancer (ALB) Ingress support on AWS EKS](#example-aws-application-load-balancer-alb-ingress-support-on-aws-eks)
```
Comment (angelinalg) [inline doc/source/cluster/kubernetes/k8s-ecosystem/ingress.md:7]:
```suggestion
  * [AWS Application Load Balancer (ALB) Ingress support on AWS EKS](#example-aws-application-load-balancer-alb-ingress-support-on-aws-eks)
```

File: doc/source/cluster/kubernetes/k8s-ecosystem/ingress.md [hunk 22]
```diff
@@ -0,0 +1,132 @@
...
+  * [Example: Manually setting up NGINX Ingress on KinD](#example-manually-setting-up-nginx-ingress-on-kind)
```
Comment (angelinalg) [inline doc/source/cluster/kubernetes/k8s-ecosystem/ingress.md:8]:
```suggestion
  * [Manually setting up NGINX Ingress on KinD](#example-manually-setting-up-nginx-ingress-on-kind)
```

File: doc/source/cluster/kubernetes/k8s-ecosystem/ingress.md [hunk 23]
```diff
@@ -0,0 +1,132 @@
...
+
+## Example: AWS Application Load Balancer (ALB) Ingress support on AWS EKS
```
Comment (angelinalg) [inline doc/source/cluster/kubernetes/k8s-ecosystem/ingress.md:10]:
```suggestion
## AWS Application Load Balancer (ALB) Ingress support on AWS EKS
```

File: doc/source/cluster/kubernetes/k8s-ecosystem/ingress.md [hunk 24]
```diff
@@ -0,0 +1,132 @@
...
+
+# Step 7: Check Ray Dashboard by ALB DNS Name. The name of the DNS Name should be like
+#        "k8s-default-rayclust-.....us-west-2.elb.amazonaws.com"
+
+# Step 8: Delete the ingress, and AWS Load Balancer controller will remove ALB.
+#        Check ALB on AWS to make sure it is removed.
+kubectl delete ingress ray-cluster-ingress
+```
+
+## Example: Manually setting up NGINX Ingress on KinD
```
Comment (angelinalg) [inline doc/source/cluster/kubernetes/k8s-ecosystem/ingress.md:74]:
```suggestion
## Manually setting up NGINX Ingress on KinD
```

File: doc/source/cluster/kubernetes/k8s-ecosystem/ingress.md [hunk 25]
```diff
@@ -0,0 +1,132 @@
...
+
+Here we provide some examples to show how to use ingress to access your Ray cluster.
+
+  * [Example: AWS Application Load Balancer (ALB) Ingress support on AWS EKS](#example-aws-application-load-balancer-alb-ingress-support-on-aws-eks)
+  * [Example: Manually setting up NGINX Ingress on KinD](#example-manually-setting-up-nginx-ingress-on-kind)
+
+## Example: AWS Application Load Balancer (ALB) Ingress support on AWS EKS
+
+### Prerequisite
+* Follow the document [Getting started with Amazon EKS – AWS Management Console and AWS CLI](https://docs.aws.amazon.com/eks/latest/userguide/getting-started-console.html#eks-configure-kubectl) to create an EKS cluster.
```
Comment (angelinalg) [inline doc/source/cluster/kubernetes/k8s-ecosystem/ingress.md:13]:
```suggestion
* Create an EKS cluster. See [Getting started with Amazon EKS – AWS Management Console and AWS CLI](https://docs.aws.amazon.com/eks/latest/userguide/getting-started-console.html#eks-configure-kubectl).
```

File: doc/source/cluster/kubernetes/k8s-ecosystem/ingress.md [hunk 26]
```diff
@@ -0,0 +1,132 @@
...
+
+* Follow the [installation instructions](https://kubernetes-sigs.github.io/aws-load-balancer-controller/latest/deploy/installation/) to set up the [AWS Load Balancer controller](https://github.com/kubernetes-sigs/aws-load-balancer-controller). Note that the repository maintains a webpage for each release. Please make sure you use the latest installation instructions.
```
Comment (angelinalg) [inline doc/source/cluster/kubernetes/k8s-ecosystem/ingress.md:15]:
```suggestion
* set up the [AWS Load Balancer controller](https://github.com/kubernetes-sigs/aws-load-balancer-controller), see [installation instructions](https://kubernetes-sigs.github.io/aws-load-balancer-controller/latest/deploy/installation/). Note that the repository maintains a webpage for each release. Confirm that you are using the latest installation instructions.
```

File: doc/source/cluster/kubernetes/k8s-ecosystem/ingress.md [hunk 27]
```diff
@@ -0,0 +1,132 @@
...
+
+* (Optional) Try [echo server example](https://github.com/kubernetes-sigs/aws-load-balancer-controller/blob/main/docs/examples/echo_server.md) in the [aws-load-balancer-controller](https://github.com/kubernetes-sigs/aws-load-balancer-controller) repository.
```
Comment (angelinalg) [inline doc/source/cluster/kubernetes/k8s-ecosystem/ingress.md:17]:
```suggestion
* (Optional) Try the [echo server example](https://github.com/kubernetes-sigs/aws-load-balancer-controller/blob/main/docs/examples/echo_server.md) in the [aws-load-balancer-controller](https://github.com/kubernetes-sigs/aws-load-balancer-controller) repository.
```

File: doc/source/cluster/kubernetes/k8s-ecosystem/ingress.md [hunk 28]
```diff
@@ -0,0 +1,132 @@
...
+
+* (Optional) Read [how-it-works.md](https://github.com/kubernetes-sigs/aws-load-balancer-controller/blob/main/docs/how-it-works.md) to understand the mechanism of [aws-load-balancer-controller](https://github.com/kubernetes-sigs/aws-load-balancer-controller).
```
Comment (angelinalg) [inline doc/source/cluster/kubernetes/k8s-ecosystem/ingress.md:19]:
```suggestion
* (Optional) Read [how-it-works.md](https://github.com/kubernetes-sigs/aws-load-balancer-controller/blob/main/docs/how-it-works.md) to understand the [aws-load-balancer-controller](https://github.com/kubernetes-sigs/aws-load-balancer-controller) mechanism.
```

File: doc/source/cluster/kubernetes/k8s-ecosystem/prometheus-grafana.md [hunk 29]
```diff
@@ -0,0 +1,359 @@
+(kuberay-prometheus-grafana)=
+
+# Prometheus & Grafana
```
Comment (angelinalg) [inline doc/source/cluster/kubernetes/k8s-ecosystem/prometheus-grafana.md:3]:
```suggestion
# Using Prometheus and Grafana
```

File: doc/source/cluster/kubernetes/k8s-ecosystem/prometheus-grafana.md [hunk 30]
```diff
@@ -0,0 +1,359 @@
...
+
+This section will describe how to monitor Ray Clusters in Kubernetes using Prometheus & Grafana.
+
+If you do not have any experience with Prometheus and Grafana on Kubernetes, I strongly recommend you watch this [YouTube playlist](https://youtube.com/playlist?list=PLy7NrYWoggjxCF3av5JKwyG7FFF9eLeL4).
```
Comment (angelinalg) [inline doc/source/cluster/kubernetes/k8s-ecosystem/prometheus-grafana.md:7]:
```suggestion
If you do not have any experience with Prometheus and Grafana on Kubernetes, watch this [YouTube playlist](https://youtube.com/playlist?list=PLy7NrYWoggjxCF3av5JKwyG7FFF9eLeL4).
```

File: doc/source/cluster/kubernetes/k8s-ecosystem/prometheus-grafana.md [hunk 31]
```diff
@@ -0,0 +1,359 @@
...
+
+## Preparation
+
+Please clone the [KubeRay repository](https://github.com/ray-project/kuberay) and checkout the `master` branch.
```
Comment (angelinalg) [inline doc/source/cluster/kubernetes/k8s-ecosystem/prometheus-grafana.md:11]:
```suggestion
Clone the [KubeRay repository](https://github.com/ray-project/kuberay) and checkout the `master` branch.
```

File: doc/source/cluster/kubernetes/k8s-ecosystem/prometheus-grafana.md [hunk 32]
```diff
@@ -0,0 +1,359 @@
...
+This tutorial requires several files in the repository.
+
+## Step 1: Create a Kubernetes cluster with Kind.
```
Comment (angelinalg) [inline doc/source/cluster/kubernetes/k8s-ecosystem/prometheus-grafana.md:14]:
```suggestion
## Step 1: Create a Kubernetes cluster with Kind.
```
```suggestion
## Step 1: Create a Kubernetes cluster with Kind
```

File: doc/source/cluster/kubernetes/k8s-ecosystem/prometheus-grafana.md [hunk 33]
```diff
@@ -0,0 +1,359 @@
...
+# ray_spill_manager_request_total{Component="raylet",NodeAddress="10.244.0.13",Type="Restored",Version="2.0.0"} 0.0
+
+# Ensure that the port (8080) for the metrics endpoint is also defined in the head's Kubernetes service.
+kubectl get service
+
+# NAME                          TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)                                         AGE
+# raycluster-kuberay-head-svc   ClusterIP   10.96.201.142   <none>        6379/TCP,8265/TCP,8080/TCP,8000/TCP,10001/TCP   106m
+```
+
+* KubeRay will expose a Prometheus metrics endpoint in port **8080** via a built-in exporter by default. Hence, we do not need to install any external exporter.
```
Comment (angelinalg) [inline doc/source/cluster/kubernetes/k8s-ecosystem/prometheus-grafana.md:84]:
```suggestion
* KubeRay exposes a Prometheus metrics endpoint in port **8080** via a built-in exporter by default. Hence, we do not need to install any external exporter.
```

File: doc/source/cluster/kubernetes/k8s-ecosystem/pyspy.md [hunk 34]
```diff
@@ -0,0 +1,77 @@
...
+- If you're running your own examples and encounter the error `Failed to write flamegraph: I/O error: No stack counts found` when viewing CPU Flame Graph, it might be due to the process being idle. Notably, using the `sleep` function can lead to this state. In such situations, py-spy filters out the idle stack traces. Refer to this [issue](https://github.com/benfred/py-spy/issues/321#issuecomment-731848950) for more information.
+
+### Step 6: Profile using Ray Dashboard
+
+- Visit http://localhost:8265/#/cluster.
+- Click `Stack Trace` for `ray::long_running_task`.
+    ![StackTrace](../images/stack_trace.png)
+- Click `CPU Flame Graph` for `ray::long_running_task`.
+    ![FlameGraph](../images/cpu_flame_graph.png)
+- For additional details on using the profiler, refer the [Ray Observability Guide](https://docs.ray.io/en/latest/ray-observability/user-guides/debug-apps/optimize-performance.html#python-cpu-profiling-in-the-dashboard).
```
Comment (angelinalg) [inline doc/source/cluster/kubernetes/k8s-ecosystem/pyspy.md:70]:
```suggestion
- For additional details on using the profiler, See [Python CPU profiling in the Dashboard](https://docs.ray.io/en/latest/ray-observability/user-guides/debug-apps/optimize-performance.html#python-cpu-profiling-in-the-dashboard).
```

**Review by angelinalg — APPROVED** [review]
Spot check of the structure looks really good. Only had time for a few copy edits. Should follow up with more. Great job, @kevin85421 !

**Comment by angelinalg** [comment]
Not being able to create child pages under Managed Kubernetes does not surprise me. Max has had to do some hacks to get around long build times that are related to large side navs, or he might've introduced limits to clean up the side nav bloat.

Re: doc tests - when you build locally, you should be running the doc tests and the CI also runs the full suite of doc tests. You don't need to do anything else.

**Comment by angelinalg** [comment]
@kevin85421 , thank you for migrating the IPYNB file to MD!

File: doc/source/_toc.yml [hunk 35]
```diff
@@ -319,43 +319,70 @@ parts:
...
-                - file: cluster/kubernetes/user-guides/logging.md
-                - file: cluster/kubernetes/user-guides/gpu.md
-                - file: cluster/kubernetes/user-guides/experimental.md
-                - file: cluster/kubernetes/user-guides/static-ray-cluster-without-kuberay.md
-            - file: cluster/kubernetes/examples
-              sections:
-                - file: cluster/kubernetes/examples/ml-example.md
-                - file: cluster/kubernetes/examples/gpu-training-example.md
-            - file: cluster/kubernetes/references
+              - file: cluster/kubernetes/getting-started
```
[the remark this answers is not in the document]
Comment (kevin85421) [reply doc/source/_toc.yml:322]: Good catch!

[the remark this answers is not in the document]
Comment (kevin85421) [reply doc/source/_toc.yml:322]: Updated https://github.com/ray-project/ray/pull/38647/commits/22730ddb7a4c5abd6c5db7d935a3cb99f0aa677e.

File: doc/source/cluster/kubernetes/examples/mobilenet-rayservice.md (hunk 2 above)
Comment (kevin85421) [reply doc/source/cluster/kubernetes/examples/mobilenet-rayservice.md:37]: Serve service here refers to Kubernetes service. Using Ray Serve service here may cause users to be confused.

File: doc/source/cluster/kubernetes/getting-started/rayjob-quick-start.md (hunk 8 above)
Comment (kevin85421) [reply doc/source/cluster/kubernetes/getting-started/rayjob-quick-start.md]: Updated https://github.com/ray-project/ray/pull/38647/commits/21dcebace5896289dcfbe067c7dae54792a0060e

File: doc/source/cluster/kubernetes/examples/mobilenet-rayservice.md (hunk 4 above)
Comment (kevin85421) [reply doc/source/cluster/kubernetes/examples/mobilenet-rayservice.md:19]: Unless you feel strongly about dividing it into two sections, I'd rather keep it as it is.

File: doc/source/cluster/kubernetes/examples/mobilenet-rayservice.md [hunk 36]
```diff
@@ -0,0 +1,47 @@
...
+* The [mobilenet.py](https://github.com/ray-project/serve_config_examples/blob/master/mobilenet/mobilenet.py) file requires `tensorflow` as a dependency. Hence, the YAML file uses `rayproject/ray-ml:2.5.0` instead of `rayproject/ray:2.5.0`.
+* `python-multipart` is required for the request parsing function `starlette.requests.form()`, so the YAML file includes `python-multipart` in the runtime environment.
+
+## Step 4: Forward the port for Ray Serve
+
+```sh
+kubectl port-forward svc/rayservice-mobilenet-serve-svc 8000
+```
+
+Note that the Serve service will be created after the Serve applications are ready and running. This process may take approximately 1 minute after all Pods in the RayCluster are running.
```
Comment (kevin85421) [inline doc/source/cluster/kubernetes/examples/mobilenet-rayservice.md:37]:
```suggestion
Note that the Serve service is created after the Ray Serve applications are ready and running. This process may take approximately 1 minute after all Pods in the RayCluster are running.
```

File: doc/source/cluster/kubernetes/examples/mobilenet-rayservice.md (hunk 2 above)
Comment (kevin85421) [reply doc/source/cluster/kubernetes/examples/mobilenet-rayservice.md:37]: https://github.com/ray-project/ray/pull/38647/files#r1303623134

**Review by zhe-thoughts — APPROVED** [review]
Yes this qualifies for merging. 

@architkulkarni could you check and give a green light of `tests-ok`? I'll wait for the green light before merging

**Comment by architkulkarni** [comment]
Failed tests:
```


//python/ray/serve:test_http_headers                                    TIMEOUT in 3 out of 3 in 60.1s
--
  | Stats over 3 runs: max = 60.1s, min = 60.1s, avg = 60.1s, dev = 0.0s
  | C:/tmp/4lhdprva/execroot/com_github_ray_project_ray/bazel-out/x64_windows-opt/testlogs/python/ray/serve/test_http_headers/test.log
  | C:/tmp/4lhdprva/execroot/com_github_ray_project_ray/bazel-out/x64_windows-opt/testlogs/python/ray/serve/test_http_headers/test_attempts/attempt_1.log
  | C:/tmp/4lhdprva/execroot/com_github_ray_project_ray/bazel-out/x64_windows-opt/testlogs/python/ray/serve/test_http_headers/test_attempts/attempt_2.log
  | //python/ray/tests:test_cancel                                          TIMEOUT in 3 out of 3 in 300.1s
  | Stats over 3 runs: max = 300.1s, min = 300.1s, avg = 300.1s, dev = 0.0s
  | C:/tmp/4lhdprva/execroot/com_github_ray_project_ray/bazel-out/x64_windows-opt/testlogs/python/ray/tests/test_cancel/test.log
  | C:/tmp/4lhdprva/execroot/com_github_ray_project_ray/bazel-out/x64_windows-opt/testlogs/python/ray/tests/test_cancel/test_attempts/attempt_1.log
  | C:/tmp/4lhdprva/execroot/com_github_ray_project_ray/bazel-out/x64_windows-opt/testlogs/python/ray/tests/test_cancel/test_attempts/attempt_2.log
  | //python/ray/serve:test_autoscaling_policy                               FAILED in 3 out of 3 in 66.9s
  | Stats over 3 runs: max = 66.9s, min = 65.4s, avg = 66.2s, dev = 0.6s
  | C:/tmp/4lhdprva/execroot/com_github_ray_project_ray/bazel-out/x64_windows-opt/testlogs/python/ray/serve/test_autoscaling_policy/test.log
  | C:/tmp/4lhdprva/execroot/com_github_ray_project_ray/bazel-out/x64_windows-opt/testlogs/python/ray/serve/test_autoscaling_policy/test_attempts/attempt_1.log
  | C:/tmp/4lhdprva/execroot/com_github_ray_project_ray/bazel-out/x64_windows-opt/testlogs/python/ray/serve/test_autoscaling_policy/test_attempts/attempt_2.log


```

Unrelated, this is a docs-only change

**Comment by kevin85421** [comment]
Hi @zhe-thoughts, I added a new commit https://github.com/ray-project/ray/pull/38647/commits/ec19d1556f06587f82b7d554ad71cfc6cfda566a, and I will let you know when the test finishes. Thanks!

**Comment by architkulkarni** [comment]
@kevin85421 when you're done updating the PR and CI is finished, you can go through the failed tests and add the tests-ok lael

**Comment by kevin85421** [comment]
`docs/readthedocs.com:anyscale-ray` builds successfully. cc @architkulkarni @zhe-thoughts

**Comment by zhe-thoughts** [comment]
Since this is a big change, @architkulkarni could you make a pass of tests and apply `tests-ok` after that? Thanks

**Comment by zhe-thoughts** [comment]
@kevin85421 sorry, need a rebase. There's conflict now

**Comment by kevin85421** [comment]
@zhe-thoughts I encountered no conflicts when rebasing with the upstream master branch locally. I still went ahead and rebased, then force-pushed. Thanks!

**Pull request merged**